symengine/polys/uintpoly.cpp

```diff
@@ -1,7 +1,245 @@
+        if (q_copy == 0_z)
+            return true;
+        return (fc % q_copy == 0_z);
+    };
+
+    while (2 * s <= T.size()) {
+        bool broken = false;
+        std::string bitmask(s, 1);
+        bitmask.resize(T.size(), 0);
+        do {
+            std::vector<unsigned int> S;
+            for (unsigned int i = 0; i < T.size(); ++i) {
+                if (bitmask[i]) {
+                    S.push_back(T[i]);
+                }
+            }
+            UIntDict G;
+
+            if (b == 1_z) {
+                integer_class q = 1_z;
+                for (auto i : S) {
+                    q *= g[i].dict_.begin()->second;
+                }
+                q %= pl;
+                if (not _test_pl(fc, q, pl))
+                    continue;
+            } else {
+                G = UIntDict(b);
+                for (auto &i : S) {
+                    G *= g[i];
+                }
+                G.itrunc(pl);
+                G.primitive(outArg(G));
+                auto q = G.dict_.begin()->second;
+                if (q != 0 and fc % q != 0)
+                    continue;
+            }
+            UIntDict H = UIntDict(b);
+            std::vector<unsigned int> T_S;
+            for (unsigned int i = 0; i < T.size(); ++i) {
+                if (not bitmask[i]) {
+                    T_S.push_back(T[i]);
+                }
+            }
+            if (b == 1_z) {
+                G = UIntDict(b);
+                for (auto &i : S) {
+                    G = G * g[i];
+                }
+                G.itrunc(pl);
+            }
+            for (auto &i : T_S) {
+                H *= g[i];
+            }
+            H.itrunc(pl);
+            integer_class G_norm = G.l1_norm();
+            integer_class H_norm = H.l1_norm();
+            if (G_norm * H_norm <= B) {
+                T = T_S;
+                G.primitive(outArg(G));
+                H.primitive(outArg(f));
+                factors.insert(
+                    UIntPoly::from_dict(get_var(), std::move(G.dict_)));
+                b = f.dict_.rbegin()->second;
+                broken = true;
+                break;
+            }
+        } while (std::prev_permutation(bitmask.begin(), bitmask.end()));
+        if (not std::next_permutation(bitmask.begin(), bitmask.end())) {
+            if (not broken)
+                s += 1;
+        }
+    }
+    if (not f.empty()) {
+        factors.insert(UIntPoly::from_dict(get_var(), std::move(f.dict_)));
+    }
+    return factors;
+}
+
+std::pair<integer_class, std::set<RCP<const UIntPoly>, RCPBasicKeyLess>>
+UIntPoly::zz_factor_sqf() const
+{
+    std::pair<integer_class, std::set<RCP<const UIntPoly>, RCPBasicKeyLess>>
+        out;
+    auto n = this->get_degree();
+    UIntDict this_dict = get_poly();
+    integer_class cont = this_dict.primitive(outArg(this_dict));
+    if (this_dict.get_lc() < 0_z) {
+        cont *= -1;
+        this_dict *= -1;
+    }
+    out.first = cont;
+    if (n == 0) {
```

Comment:
@nishnik , I am currently going with implementing wrappers for FLINT rather than our own factorisation(Which I will take up after gsoc ) as per the [discussion](https://groups.google.com/forum/#!topic/symengine/vRKtOR_tnhI) in the mailing list.